OCPBUGSM-16391 - BE return all-hosts-are-ready-to-install=success whe…

gamli75 opened the pull request:
…n no host connected

@filanov @avishayt @ori-amizur please review

filanov commented on internal/cluster/validator.go:
```diff
@@ -281,6 +281,9 @@ func (v *clusterValidator) printSufficientMastersCount(context *clusterPreproces
 }
 
 func (v *clusterValidator) allHostsAreReadyToInstall(c *clusterPreprocessContext) validationStatus {
+	if len(c.cluster.Hosts) < common.MinMasterHostsNeededForInstallation {
```
does it count disabled hosts?

gamli75 replied:
fixed

ori-amizur commented on internal/cluster/validator.go:
```diff
@@ -281,6 +281,9 @@ func (v *clusterValidator) printSufficientMastersCount(context *clusterPreproces
 }
 
 func (v *clusterValidator) allHostsAreReadyToInstall(c *clusterPreprocessContext) validationStatus {
+	if len(c.cluster.Hosts) < common.MinMasterHostsNeededForInstallation {
```
Isn't it now a duplication of sufficient-master-count validation? Why do we need two validations that check the same thing?

gamli75 replied:
no, there are a few scenarios: 
1. 3 masters in a known state and one worker with insufficient - sufficient-master-count validation= true allHostsAreReadyToInstall= false
2. 2 masters & 2 workers are in a known state - sufficient-master-count validation= false, allHostsAreReadyToInstall= true

ori-amizur commented on internal/cluster/validator.go:
```diff
@@ -294,18 +294,28 @@ func (v *clusterValidator) printSufficientMastersCount(context *clusterPreproces
+
 		}
 	}
+	if numberOfEnabledHosts < common.MinMasterHostsNeededForInstallation {
```
Usually, we use pending validation to indicate that this validation depends on other validation.  This is not the case here. The UI hides pending validations.  
So the pending status should be removed.
The same information is received by "sufficient master count" and "all hosts ready to install"

gamli75 replied:
@ori-amizur as you can see in the Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1872668 @avishayt  agreed that it should be pending.

ori-amizur replied:
So this validation will be hidden by UI in this case.  I think it is wrong.
A better solution will be to display a different validation message in this case and the result should be failure.

slaviered replied:
numberOfEnabledHosts contains also workers so it can not be compared to MinMasterHostNeededForInstallation

gamli75 replied:
there was a bug that if we have only 2 known hosts in the cluster - the validation of "All hosts are ready to install" return true instead of false - the role of the host is irrelevant if we have only 1 or 2 hosts that are registered and in known state

gamli75 commented:
/retest

gamli75 commented:
@filanov @ori-amizur please review again
I fixed 2 bugs and added tests:
https://bugzilla.redhat.com/show_bug.cgi?id=1872668
https://bugzilla.redhat.com/show_bug.cgi?id=1879383

filanov commented on internal/cluster/cluster_test.go:
```diff
@@ -144,13 +144,13 @@ var _ = Describe("cluster monitor", func() {
 			Expect(db.Create(&c).Error).ShouldNot(HaveOccurred())
 			Expect(err).ShouldNot(HaveOccurred())
 			mockMetric.EXPECT().ClusterInstallationFinished(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+			mockHostAPI.EXPECT().IsValidMasterCandidate(gomock.Any(), gomock.Any(), gomock.Any()).Return(true, nil).AnyTimes()
```
why any time?

filanov commented on internal/cluster/transition_test.go:
```diff
@@ -351,6 +351,11 @@ var _ = Describe("Refresh Cluster - No DHCP", func() {
 	}
+
+	mockHostAPIIsValidMasterCandidateTrue := func() {
+		mockHostAPI.EXPECT().IsValidMasterCandidate(gomock.Any(), gomock.Any(), gomock.Any()).Return(true, nil).AnyTimes()
```
why to use anytime?

filanov commented on internal/cluster/transition_test.go:
```diff
@@ -1504,6 +1573,9 @@ var _ = Describe("Refresh Cluster - With DHCP", func() {
 					cluster.MachineNetworkCidrUpdatedAt = time.Now().Add(-3 * time.Minute)
 				}
+				if t.srcState == models.ClusterStatusInstalled {
+					cluster.ValidationsInfo = getSuccessfulValidationInfo()
```
Why do you need it? it's not clear why do you overwrite the validations info?

gamli75 replied:
I added a comment

slaviered commented on internal/cluster/validator.go:
```diff
@@ -341,7 +369,7 @@ func (v *clusterValidator) printAllHostsAreReadyToInstall(context *clusterPrepro
 		return "All hosts in the cluster are ready to install."
 	case ValidationFailure:
-		return "The cluster has hosts that are not ready to install."
+		return "The cluster has hosts that are not ready to install or there are not enough discovered hosts."
```
shouldn't this be covered in the sufficient host count validation?

slaviered commented on internal/cluster/mock_cluster_api.go:
```diff
@@ -6,13 +6,12 @@ package cluster
 	gorm "github.com/jinzhu/gorm"
 	common "github.com/openshift/assisted-service/internal/common"
 	s3wrapper "github.com/openshift/assisted-service/pkg/s3wrapper"
+	reflect "reflect"
```
you have a lot of extra mocks that do not relates to your change. run make format to get rid of them

gamli75 commented:
closed this because one of the tickets was closed - I'll open a new PR for the specific bug
